synapse/storage/databases/main/deviceinbox.py

```diff
@@ -733,74 +680,94 @@ async def _remove_hidden_devices_from_device_inbox(
         Returns:
             The number of deleted rows
         """
+        number_deleted, finished = await self.db_pool.runInteraction(
+            "_remove_hidden_devices_from_device_inbox_txn",
+            self._remove_devices_from_device_inbox_txn,
+            self.REMOVE_HIDDEN_DEVICES,
+            progress,
+            batch_size,
+        )
 
-        def _remove_hidden_devices_from_device_inbox_txn(
-            txn: LoggingTransaction,
-        ) -> int:
-            """stream_id is not unique
-            we need to use an inclusive `stream_id >= ?` clause,
-            since we might not have deleted all hidden device messages for the stream_id
-            returned from the previous query
-
-            Then delete only rows matching the `(user_id, device_id, stream_id)` tuple,
-            to avoid problems of deleting a large number of rows all at once
-            due to a single device having lots of device messages.
-            """
+        # The task is finished when no more lines are deleted.
+        if finished:
+            await self.db_pool.updates._end_background_update(
+                self.REMOVE_HIDDEN_DEVICES
+            )
 
-            last_stream_id = progress.get("stream_id", 0)
+        return number_deleted
 
-            sql = """
-                SELECT device_id, user_id, stream_id
-                FROM device_inbox
-                WHERE
-                    stream_id >= ?
-                    AND (device_id, user_id) IN (
-                        SELECT device_id, user_id FROM devices WHERE hidden = ?
-                    )
-                ORDER BY stream_id
-                LIMIT ?
-            """
+    def _remove_devices_from_device_inbox_txn(
```

Comment:
I've tried to be a bit clever here and have both updates call the same function since they're so similar, lmk if it's more desirable to have each run its own nested function

Reply:
per my comment - I think we can actually combine the two updates into one query relatively easily.